torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -336,6 +336,33 @@ void Reducer::check_grad_layout(
   }
 }
 
+void Reducer::point_grads_to_bucket() {
+  std::unordered_set<int> zeroed_buckets;
+  for (size_t variable_index = 0; variable_index < params_.size();
+       ++variable_index) {
+    const auto& bucket_index = variable_locators_[variable_index];
+    auto& bucket = buckets_[bucket_index.bucket_index];
+    if (zeroed_buckets.find(bucket_index.bucket_index) ==
+        zeroed_buckets.end()) {
+      bucket.gradients.zero_();
+      zeroed_buckets.insert(bucket_index.bucket_index);
+    }
+    auto& variable = bucket.variables[bucket_index.intra_bucket_index];
+    auto& bucket_view = bucket.bucket_views_out[bucket_index.intra_bucket_index];
+
+    runGradCallbackForVariable(variable, [&](auto& grad) {
+      if (!grad.defined()) {
+        // grad is None: assuming user called zero_grad(set_to_none=True)
+        // point grad to bucket
+        grad = bucket_view;
```

Comment:
C++ noob question: Will the `bucket_view` reference be bound to the anonymous grad callback's closure automatically (even though the callback is only defined to take in the `grad` reference)?

Reply:
Also, does assigning a reference to a reference persist? Are we sure that the `grad` is being set to `bucket_view` even after this lambda returns?

Reply:
In the lambda, the stuff in `[...]` defines the variables to capture, and `[&]` means to capture the surrounding context by reference. `(auto& grad)` is the argument taken into this callback.

I believe this will work as we do the same thing here: https://fburl.com/code/pshiibxd,